### PR TITLE
mgr: use un-deprecated APIs to initialize Python interpretor

### PR DESCRIPTION
**Pull request by tchaikov**
`Py_SetProgramName()`, `PySys_SetArgv()` and `Py_GetPath()` were deprecated since CPython 3.11, see https://docs.python.org/3/c-api/init_config.html . `Py_InitializeFromConfig()` and friends were introduced by CPython 3.8,
but we still need to support CPython 3.6 which is shipped by CentOS8. so we have to be backward compatible with the older Python versions.

so let's use new machinary to initialize the Python interpretor if the tree is compiled with CPython 3.8 and up, i.e., `PY_VERSION_HEX >= 0x03080000`. so that this piece of code can be tested on ubuntu:jammy, which ships Python 3.10, see https://packages.ubuntu.com/jammy/amd64/python3

this change addresses following compiling warning:

```
[428/753] Building CXX object src/mgr/CMakeFiles/ceph-mgr.dir/PyModuleRegistry.cc.o
/var/ssd/ceph/src/mgr/PyModuleRegistry.cc: In member function ‘void PyModuleRegistry::init()’:
/var/ssd/ceph/src/mgr/PyModuleRegistry.cc:49:20: warning: ‘void Py_SetProgramName(const wchar_t*)’ is deprecated [-Wdeprecated-declarations]
   49 |   Py_SetProgramName(const_cast<wchar_t*>(WCHAR(MGR_PYTHON_EXECUTABLE)));
      |   ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/python3.12/Python.h:94,
                 from /var/ssd/ceph/src/mgr/PyModule.h:22,
                 from /var/ssd/ceph/src/mgr/PyModuleRegistry.h:18,
                 from /var/ssd/ceph/src/mgr/PyModuleRegistry.cc:14:
/usr/include/python3.12/pylifecycle.h:37:38: note: declared here
   37 | Py_DEPRECATED(3.11) PyAPI_FUNC(void) Py_SetProgramName(const wchar_t *);
      |                                      ^~~~~~~~~~~~~~~~~`
```

### CODE REVIEW

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by ronen-fr** [comment]
PySys_SetArgv() seems to appear twice in the PR description

File: src/mgr/PyModuleRegistry.cc [hunk 1]
```diff
@@ -46,14 +47,56 @@ void PyModuleRegistry::init()
...
   // Set up global python interpreter
 #define WCHAR(s) L ## #s
+#if PY_VERSION_HEX >= 0x03080000
+  PyConfig py_config;
+  PyConfig_InitIsolatedConfig(&py_config);
+  BOOST_SCOPE_EXIT_ALL(&py_config) {
+    PyConfig_Clear(&py_config);
+  };
+  // site_import is 1 by default, but set it explicitly as we do import
+  // site packages manually for Python < 3.8
```
Comment (ronen-fr) [inline src/mgr/PyModuleRegistry.cc:57]: I am not sure I understand the comment. What's the relevance of what we do in the <3.8 case?

File: src/mgr/PyModuleRegistry.cc [hunk 2]
```diff
@@ -46,14 +47,56 @@ void PyModuleRegistry::init()
...
+  const wchar_t* argv[] = {L"ceph-mgr"};
+  PyConfig_SetArgv(&py_config, 1, (wchar_t *const *)argv);
+  // Add more modules
+  if (g_conf().get_val<bool>("daemonize")) {
+    PyImport_AppendInittab("ceph_logger", PyModule::init_ceph_logger);
+  }
+  PyImport_AppendInittab("ceph_module", PyModule::init_ceph_module);
+  // Configure sys.path to include mgr_module_path
+  const auto mgr_module_path = g_conf().get_val<std::string>("mgr_module_path");
+  std::wstring module_search_paths(std::begin(mgr_module_path), std::end(mgr_module_path));
```
Comment (ronen-fr) [inline src/mgr/PyModuleRegistry.cc:74]: should we assert if mgr_module_path isn't there?

**Review by ronen-fr — APPROVED** [review]

File: src/mgr/PyModuleRegistry.cc (hunk 1 above)
Comment (tchaikov) [reply src/mgr/PyModuleRegistry.cc:57]: i wanted to explain why i was adding a noop. the goal is to mirror the behavior of another branch which is picked by the preprocessor.

**Comment by tchaikov** [comment]
> PySys_SetArgv() seems to appear twice in the PR description

thank you for pointing this out. fixed.

**Comment by tchaikov** [comment]
marking as a draft, as following tests are failing
```
The following tests FAILED:
	 36 - smoke.sh (Timeout)
	197 - safe-to-destroy.sh (Timeout)
```

File: src/mgr/PyModuleRegistry.cc (hunk 2 above)
Comment (tchaikov) [reply src/mgr/PyModuleRegistry.cc:74]: isn't where? in `py_config.module_search_paths`?

**Comment by tchaikov** [comment]
changelog

- do not enable isolated mode, otherwise it'd be difficult to have access to site packages.
- customize pyconfig to make it more isolated.
- set `PYTHONPATH` instead of appending to `module_search_paths`, because `module_search_paths` starts as an empty list. so we are not be able to access site packages if we only append our own paths to it.
- use `ceph_assertf()` for printing the details of failure.

File: src/mgr/PyModuleRegistry.cc (hunk 2 above)
Comment (ronen-fr) [reply src/mgr/PyModuleRegistry.cc]: if the configuration option 'mgr_module_path' is not set

Comment (tchaikov) [reply src/mgr/PyModuleRegistry.cc]: i see. but the default value this option is not empty, see https://github.com/ceph/ceph/blob/6847404eda0fe207e6991561eaf3df91aa0c8edf/src/common/options/mgr.yaml.in#L111 . if we consider it a bug, and really want to handle this case, i'd suggest create tracker ticket. and to implement it in another PR. as it is an preexisting behavior, and not in the scope of this change, which is created to address the deprecated Python APIs used in ceph-mgr. 

what do you think?

Comment (ronen-fr) [reply src/mgr/PyModuleRegistry.cc]: I agree it's not a part of this PR.

**Comment by tchaikov** [comment]
@yuriw hi Yuri, i see you added and then removed wip-yuri6-testing label. did this test fail the test or you just decided to remove it from that batch for the reason like lowering the risk?

**Comment by yuriw** [comment]
@tchaikov see this trello https://trello.com/c/PB1LsGKm

**Comment by ljflores** [comment]
Hey @tchaikov, I suspected this PR for breaking tests. I wanted Yuri to run the tests w/o this PR to confirm though. I'll post here exactly the symptoms if your PR is confirmed to be the cause.

**Comment by tchaikov** [comment]
@ljflores @yuriw hi Laura and Yuri, thanks!

**Comment by NitzanMordhai** [comment]
Rados approved: https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrellocomcPB1LsGKm1970-wip-yuri8-testing-2024-03-11-1138-old-wip-yuri8-testing-2024-03-06-1329-old-wip-yuri8-testing-2024-03-04-1226-old-wip-yuri8-test

**Comment by tchaikov** [comment]
@ljflores @yuriw hi Laura and Yuri, did you confirm that this change is the offending change? all changes except this one tested the above batch have been merged. so i guess this changeset should also be merged, am i right?

**Comment by tchaikov** [comment]
i will merge this change in two days if no objections.

**Comment by tchaikov** [comment]
resurrected at #58199

**Pull request merged**